src/transport/SecureSessionMgr.cpp

```diff
@@ -292,12 +317,47 @@ void SecureSessionMgr::OnMessageReceived(const PeerAddress & peerAddress, System
 void SecureSessionMgr::MessageDispatch(const PacketHeader & packetHeader, const Transport::PeerAddress & peerAddress,
                                        System::PacketBufferHandle && msg)
 {
+    Transport::UnauthenticatedSession * session = mUnauthenticatedSessions.FindOrAllocateEntry(peerAddress);
+    if (session == nullptr)
+    {
+        ChipLogError(Inet, "UnauthenticatedSession exhausted");
+        return;
+    }
+
+    SecureSessionMgrDelegate::DuplicateMessage isDuplicate = SecureSessionMgrDelegate::DuplicateMessage::No;
+
+    // Verify message counter
+    CHIP_ERROR err = session->GetPeerMessageCounter().VerifyOrTrustFirst(packetHeader.GetMessageId());
+    if (err == CHIP_ERROR_DUPLICATE_MESSAGE_RECEIVED)
+    {
+        ChipLogDetail(Inet, "Received a duplicate message with MessageId: %" PRIu32, packetHeader.GetMessageId());
+        isDuplicate = SecureSessionMgrDelegate::DuplicateMessage::Yes;
+        err         = CHIP_NO_ERROR;
+    }
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(Inet, "Message counter verify failed, err = %" CHIP_ERROR_FORMAT, err.Format());
+        return;
+    }
+
+    mUnauthenticatedSessions.MarkSessionActive(*session);
+
+    PayloadHeader payloadHeader;
+    ReturnOnFailure(payloadHeader.DecodeAndConsume(msg));
+
+    if (isDuplicate == SecureSessionMgrDelegate::DuplicateMessage::Yes && !payloadHeader.NeedsAck())
+    {
+        // If it's a duplicate message, but doesn't require an ack, let's drop it right here to save CPU
+        // cycles on further message processing.
+        return;
+    }
```

Comment:
I don't think this is a very useful optimization.  For one thing, it costs flash (albeit not much).  So we save a tiny bit of CPU at the cost of a tiny bit of flash.

I would rather be confident that `OnMessageReceived` will do the right thing in all of the four possible cases for isDuplicate and NeedsAck.  This increases fragility because `OnMessageReceived` should support the isDuplicate=false, NeedsAck=false case, but may not.  With this optimization, we would never know that.

```suggestion
```

Reply:
Applied